README.md
# Chess Embeddings

Pre-trained [GloVe](git@github.com:stanfordnlp/GloVe.git) word embeddings created from ~3.5 million chess games from the [ChessDB dataset](https://chessdb.sourceforge.net/).

Embeddings were trained on ~3GB of convenient PGN-formatted data made available via [this dataset](http://chess-research-project.readthedocs.io/en/latest/) (mirror download also available [here](https://brannondorsey.s3.amazonaws.com/ChessEmbeddings/all.pgn.zip)). You can find these embeddings inside `data/embeddings/`. I've also provided scripts and instructions to easily create your own embeddings using more data from the above dataset if you prefer.

## Downloads

* [Pre-trained Chess embeddings](https://brannondorsey.s3.amazonaws.com/ChessEmbeddings/embeddings.zip) (215 MB)
* [Text file used to train these embeddings](https://brannondorsey.s3.amazonaws.com/ChessEmbeddings/moves_from_3561458_games.zip) (374 MB)
* [Original ChessDB PGN dataset](https://brannondorsey.s3.amazonaws.com/ChessEmbeddings/all.pgn.zip) (830 MB)
* [Original ChessDB PGN dataset transformed to CSV format](https://brannondorsey.s3.amazonaws.com/ChessEmbeddings/all.csv.zip) (540 MB)
* [Python pickle of all into a `pgnparser.PGNGame` array](https://brannondorsey.s3.amazonaws.com/ChessEmbeddings/all.pickle.zip) (requires ~32GB of RAM to load) (656 MB)

## Creating Your Own Embeddings

Start by cloning and building [Stanford's GloVe repo](https://github.com/stanfordnlp/GloVe). I've provided a script that does this for you with:

```bash
./setup.sh
```

The GloVe tools require a file filled with words (in our case chess moves) seperated by spaces to train the embedding vectors with. Because GloVe trains word vectors that represent word colocation statistics, the order of the words in this input file is very important. 

I've included the file that I used to generate the embeddings included with this repo (created from the above chess dataset) in `data/train_moves.txt`. This file contains the first 10,108,811 moves in the chess dataset. You will also find a `data/test_moves.txt` with 4,335,426 moves that the word vectors have not been trained with. In both of these files, games are not explicitly seperated, however doing so is as easy as splitting on moves `0-1`, `1-0`, or `1/2-1/2`. 

If you would like to create embeddings using more moves from the dataset (and there are certaintly a **lot** more moves), you must create a file similar to `data/train_moves.txt`. I intend to provide a script for doing this as well, but haven't gotten around to it yet (see `scripts` for a messy file with bits and pieces that might help you. Here be dragons.)

Once this is done, open `create_embeddings.sh` and edit the value of the `WORD_FILE` variable to point towards your word file.

There are many other variables here that you can change to customize the training of your GloVe vectors. Once you've edited the appropriate variables, create the embeddings with:

```bash
./create_embeddings.sh
```

Your new embeddings, by default, will override the existing files in `data/embeddings`

## Analyzing the Data

I've provided a few Jupyter notebooks that analyze/visualize the moves in `data/train_moves.txt` in different ways. Poke around `notebooks` to learn some interesting things about chess :) 


CHANGELOG.md
# Changelog

## 2.0

* Train all embeddings for 1,000 iterations, instead of 200 used in 1.0.
* Replace use of `1-0`, `0-1`, and `1/2-1/2` as delimiters of games in GloVe training set with `<start>` and `<end>` tokens. The result of each game is no longer inherently preserved or learned during the embedding process.
* Added power-of-two versions of pre-trained embeddings in addition to the sizes used in the original GloVe paper. You will now find embeddings for 8, 16, 32, 64, 128, and 512 embeddings in [`data/embeddings/`](data/embeddings) in addition to the previous dimensions (10, 25, 50, 100, 200, 300).
* Train embeddings on full 3.5 million game corpus instead of the 100,000+ game sample used in 1.0. That's 2.9 GB of data instead of 40 MB.
* Created an optimized version of `load_parse_and_pickle()` which uses `multiprocessing` and a new[`TaskManager`](src/TaskManager.py) class that implements an efficient consumer/producer concurrency model. See [`scripts/load_parse_and_pickle.py`](scripts/load_parse_and_pickle_games.py) for details.
* Add `requirements.txt` for reproducibility
* Update GloVe submodule to latest `master`
* Add `scripts/pgn_to_csv.py` to convert `all.pgn` to a CSV for easier parsing/manipulation in downstream tasks.

## 1.0

* Original 2017 release. See this tree for a [snapshot](https://github.com/brannondorsey/ChessEmbeddings/tree/e521638b39ea4af1efa4c62ab519406324fea385) of the original 1.0 release.


create_embeddings.sh
#!/bin/bash

# Last run time: ./create_embeddings.sh  373374.72s user 1456.40s system 1148% cpu 9:04:07.97 total
WORD_FILE=data/moves_from_3561458_games.txt
OUT_DIR=data/embeddings

MIN_COUNT=5
WINDOW_SIZE=15
# Includes the original vector sizes from the GloVe paper + their power of two equivalents
DIMENSIONS=( 8 10 16 25 32 50 64 100 128 200 256 300 512 )
ITERATIONS=1000
NUM_THREADS=12

mkdir -p $OUT_DIR/

GloVe/build/vocab_count -min-count $MIN_COUNT -verbose 2 < $WORD_FILE > "$OUT_DIR/vocab.txt"
GloVe/build/cooccur -memory 4.0 -vocab-file "$OUT_DIR/vocab.txt" -verbose 2 -window-size $WINDOW_SIZE < $WORD_FILE > "$OUT_DIR/cooccurrence.bin"
GloVe/build/shuffle -memory 4.0 -verbose 2 < "$OUT_DIR/cooccurrence.bin" > "$OUT_DIR/cooccurrence.shuf.bin"

for D in "${DIMENSIONS[@]}"
do
	GloVe/build/glove -save-file "$OUT_DIR/vectors_d$D" -threads $NUM_THREADS \
					  -input-file "$OUT_DIR/cooccurrence.shuf.bin" \
				      -x-max 10 -iter $ITERATIONS -vector-size $D \
				      -binary 2 -vocab-file "$OUT_DIR/vocab.txt" -verbose 2
done

rm $OUT_DIR/cooccurrence.bin
rm $OUT_DIR/cooccurrence.shuf.bin

#save these settings to file
echo "WORD_FILE=$WORD_FILE"                      >  $OUT_DIR/settings.txt
echo "OUT_DIR=$OUT_DIR"                          >> $OUT_DIR/settings.txt
echo "MIN_COUNT=$MIN_COUNT"                      >> $OUT_DIR/settings.txt
echo "WINDOW_SIZE=$WINDOW_SIZE"                  >> $OUT_DIR/settings.txt
echo "DIMENSIONS=$(echo ${DIMENSIONS[*]// /|})"  >> $OUT_DIR/settings.txt
echo "ITERATIONS=$ITERATIONS"                    >> $OUT_DIR/settings.txt

